WalletWasabi/WabiSabi/Client/ArenaClient.cs

```diff
@@ -0,0 +1,99 @@
+
+		public ValueTask<Guid> RegisterInputAsync(Money amount, OutPoint outPoint, Key key, Guid roundId, uint256 roundHash) =>
+			RegisterInputAsync(
+				new[] { amount },
+				new[] { outPoint },
+				new[] { key },
+				roundId,
+				roundHash);
+
+		public async ValueTask<Guid> RegisterInputAsync(
+			IEnumerable<Money> amounts,
+			IEnumerable<OutPoint> outPoints,
+			IEnumerable<Key> keys,
+			Guid roundId,
+			uint256 roundHash)
+		{
+			static byte[] GenerateOwnershipProof(Key key, uint256 roundHash) => OwnershipProof.GenerateCoinJoinInputProof(
+				key,
+				new CoinJoinInputCommitmentData("CoinJoinCoordinatorIdentifier", roundHash)).ToBytes();
+
+			var registrableInputs = outPoints
+				.Zip(keys, (outPoint, key) => (outPoint, key))
+				.Select(x => new InputRoundSignaturePair(x.outPoint, GenerateOwnershipProof(x.key, roundHash)));
+
+			var (zeroAmountCredentialRequest, zeroAmountCredentialResponseValidation) = AmountCredentialClient.CreateRequestForZeroAmount();
+			var (zeroWeightCredentialRequest, zeroWeightCredentialResponseValidation) = WeightCredentialClient.CreateRequestForZeroAmount();
+
+			var inputRegistrationResponse = await RequestHandler.RegisterInputAsync(
+				new InputsRegistrationRequest(
+					roundId,
+					registrableInputs,
+					zeroAmountCredentialRequest,
+					zeroWeightCredentialRequest)).ConfigureAwait(false);
+
+			AmountCredentialClient.HandleResponse(inputRegistrationResponse.AmountCredentials, zeroAmountCredentialResponseValidation);
+			WeightCredentialClient.HandleResponse(inputRegistrationResponse.WeightCredentials, zeroWeightCredentialResponseValidation);
+
+			return inputRegistrationResponse.AliceId;
+		}
+
+		public async Task ConfirmConnectionAsync(Guid roundId, Guid aliceId, IEnumerable<Credential> credentialsToPresent, IEnumerable<Money> newAmount)
+		{
+			var (realAmountCredentialRequest, realAmountCredentialResponseValidation) = AmountCredentialClient.CreateRequest(
+				newAmount.Select(x => x.Satoshi),
+				credentialsToPresent);
+
+			var inputsCount = 1; // ?
+			var (realWeightCredentialRequest, realWeightCredentialResponseValidation) = WeightCredentialClient.CreateRequest(
+				new[] { 1_000L - (inputsCount) * 4 * Constants.P2wpkhInputVirtualSize },
+				WeightCredentialClient.Credentials.ZeroValue.Take(2));
+
+			var (zeroAmountCredentialRequest, zeroAmountCredentialResponseValidation) = AmountCredentialClient.CreateRequestForZeroAmount();
+			var (zeroWeightCredentialRequest, zeroWeightCredentialResponseValidation) = WeightCredentialClient.CreateRequestForZeroAmount();
+
+			var confirmConnectionResponse = await RequestHandler.ConfirmConnectionAsync(
+				new ConnectionConfirmationRequest(
+					roundId,
+					aliceId,
+					zeroAmountCredentialRequest,
+					realAmountCredentialRequest,
+					zeroWeightCredentialRequest,
+					realWeightCredentialRequest)).ConfigureAwait(false);
+
+			if (confirmConnectionResponse is { RealAmountCredentials: { }, RealWeightCredentials: { } })
```

Comment:
What is happening here? Is this a null check? Hard to read for me.

Reply:
It is using pattern matching to check if the `confirmConnectionResponse` variable is an object with the two properties no null.
  
The alternative is more verbose:

```c#
if (confirmConnectionResponse.RealAmountCredentials is { } && confirmConnectionResponse.RealWeightCredentials is { })

```

Reply:
maybe a new method for `ConnectionConfirmationResponse`? `IsConfirmed` perhaps?

Reply:
That would be a calculated property only used here.

Reply:
@molnard I will merge using my superpowers. If you don't like this line and prefer the alternative version just let me know and I will change it in the next PR (or you can change it in your PR, it is okay for me)

Reply:
> That would be a calculated property only used here.

it's not about reuse but readability, what the client cares about is whether it got real credentials, but this is not very obvious as a signalling mechanism for readyness. i think it would be clearer if the code conveyed that the confirmation of the input was accepted and input registration has concluded.

Reply:
It is OK I will live with it.